File: tdf-hq-ui/src/api/courses.ts
import { del, get, patch, post } from './client';
import type { components } from './generated/types';

export type CourseMetadata = components['schemas']['CourseMetadata'] & {
  remaining?: number | null;
  sessionStartHour?: number | null;
  sessionDurationHours?: number | null;
  instructorName?: string | null;
  instructorBio?: string | null;
  instructorAvatarUrl?: string | null;
};
export type CourseRegistrationRequest = components['schemas']['CourseRegistrationRequest'];
export type CourseRegistrationResponse = components['schemas']['CourseRegistrationResponse'];
export type CourseRegistrationStatusUpdate = components['schemas']['CourseRegistrationStatusUpdate'];

export interface CourseSessionIn {
  label: string;
  date: string;
  order?: number | null;
}

export interface CourseSyllabusIn {
  title: string;
  topics: string[];
  order?: number | null;
}

export interface CourseUpsert {
  slug: string;
  title: string;
  subtitle?: string | null;
  format?: string | null;
  duration?: string | null;
  priceCents: number;
  currency: string;
  capacity: number;
  sessionStartHour?: number | null;
  sessionDurationHours?: number | null;
  locationLabel?: string | null;
  locationMapUrl?: string | null;
  whatsappCtaUrl?: string | null;
  landingUrl?: string | null;
  daws: string[];
  includes: string[];
  instructorName?: string | null;
  instructorBio?: string | null;
  instructorAvatarUrl?: string | null;
  sessions: CourseSessionIn[];
  syllabus: CourseSyllabusIn[];
}

export interface CourseRegistrationDTO {
  crId: number;
  crCourseSlug: string;
  crPartyId?: number | null;
  crFullName?: string | null;
  crEmail?: string | null;
  crPhoneE164?: string | null;
  crSource: string | null;
  crStatus: string;
  crAdminNotes?: string | null;
  crHowHeard?: string | null;
  crUtmSource?: string | null;
  crUtmMedium?: string | null;
  crUtmCampaign?: string | null;
  crUtmContent?: string | null;
  crCreatedAt: string;
  crUpdatedAt: string;
}

export interface CourseRegistrationReceiptDTO {
  crrId: number;
  crrRegistrationId: number;
  crrPartyId?: number | null;
  crrFileUrl: string;
  crrFileName?: string | null;
  crrMimeType?: string | null;
  crrNotes?: string | null;
  crrUploadedBy?: number | null;
  crrCreatedAt: string;
  crrUpdatedAt: string;
}

export interface CourseRegistrationFollowUpDTO {
  crfId: number;
  crfRegistrationId: number;
  crfPartyId?: number | null;
  crfEntryType: string;
  crfSubject?: string | null;
  crfNotes: string;
  crfAttachmentUrl?: string | null;
  crfAttachmentName?: string | null;
  crfNextFollowUpAt?: string | null;
  crfCreatedBy?: number | null;
  crfCreatedAt: string;
  crfUpdatedAt: string;
}

export interface CourseRegistrationDossierDTO {
  crdRegistration: CourseRegistrationDTO;
  crdReceipts: CourseRegistrationReceiptDTO[];
  crdFollowUps: CourseRegistrationFollowUpDTO[];
  crdCanMarkPaid: boolean;
}

export interface CourseRegistrationNotesUpdate {
  notes?: string | null;
}

export interface CourseRegistrationReceiptCreate {
  fileUrl: string;
  fileName?: string | null;
  mimeType?: string | null;
  notes?: string | null;
}

export interface CourseRegistrationReceiptUpdate {
  fileUrl?: string | null;
  fileName?: string | null;
  mimeType?: string | null;
  notes?: string | null;
}

export interface CourseRegistrationFollowUpCreate {
  entryType?: string | null;
  subject?: string | null;
  notes: string;
  attachmentUrl?: string | null;
  attachmentName?: string | null;
  nextFollowUpAt?: string | null;
}

export interface CourseRegistrationFollowUpUpdate {
  entryType?: string | null;
  subject?: string | null;
  notes?: string | null;
  attachmentUrl?: string | null;
  attachmentName?: string | null;
  nextFollowUpAt?: string | null;
}

export interface CourseEmailEventDTO {
  ceId: number;
  ceCourseSlug: string;
  ceRegistrationId?: number | null;
  ceRecipientEmail: string;
  ceRecipientName?: string | null;
  ceEventType: string;
  ceStatus: string;
  ceMessage?: string | null;
  ceCreatedAt: string;
}

export interface CourseCohortOptionDTO {
  ccSlug: string;
  ccTitle?: string | null;
}

const courseBase = (slug: string) => `/public/courses/${slug}`;

const requirePositiveInteger = (value: number, field: string): number => {
  if (!Number.isSafeInteger(value) || value <= 0) {
    throw new Error(`${field} debe ser un entero positivo.`);
  }
  return value;
};

const normalizeCourseSlug = (slug: string): string => {
  const trimmed = slug.trim();
  if (trimmed === '') {
    throw new Error('slug no puede estar vacío.');
  }
  return encodeURIComponent(trimmed);
};

const setTrimmedParam = (search: URLSearchParams, key: string, value?: string): void => {
  const trimmed = value?.trim();
  if (trimmed) search.set(key, trimmed);
};

export const Courses = {
  upsert: (payload: CourseUpsert) => post<CourseMetadata>('/admin/courses', payload),
  listCohorts: () => get<CourseCohortOptionDTO[]>('/admin/courses/cohorts'),
  getMetadata: (slug: string) => get<CourseMetadata>(courseBase(normalizeCourseSlug(slug))),
  register: (slug: string, payload: CourseRegistrationRequest) =>
    post<CourseRegistrationResponse>(`${courseBase(normalizeCourseSlug(slug))}/registrations`, payload),
  updateStatus: (slug: string, registrationId: number, payload: CourseRegistrationStatusUpdate) =>
    patch<CourseRegistrationResponse>(
      `/admin/courses/${normalizeCourseSlug(slug)}/registrations/${requirePositiveInteger(registrationId, 'registrationId')}/status`,
      payload,
    ),
  getRegistrationDossier: (slug: string, registrationId: number) =>
    get<CourseRegistrationDossierDTO>(
      `/admin/courses/${normalizeCourseSlug(slug)}/registrations/${requirePositiveInteger(registrationId, 'registrationId')}/dossier`,
    ),
  updateRegistrationNotes: (slug: string, registrationId: number, payload: CourseRegistrationNotesUpdate) =>
    patch<CourseRegistrationDTO>(
      `/admin/courses/${normalizeCourseSlug(slug)}/registrations/${requirePositiveInteger(registrationId, 'registrationId')}/notes`,
      payload,
    ),
  createReceipt: (slug: string, registrationId: number, payload: CourseRegistrationReceiptCreate) =>
    post<CourseRegistrationReceiptDTO>(
      `/admin/courses/${normalizeCourseSlug(slug)}/registrations/${requirePositiveInteger(registrationId, 'registrationId')}/receipts`,
      payload,
    ),
  updateReceipt: (
    slug: string,
    registrationId: number,
    receiptId: number,
    payload: CourseRegistrationReceiptUpdate,
  ) =>
    patch<CourseRegistrationReceiptDTO>(
      `/admin/courses/${normalizeCourseSlug(slug)}/registrations/${requirePositiveInteger(registrationId, 'registrationId')}/receipts/${requirePositiveInteger(receiptId, 'receiptId')}`,
      payload,
    ),
  deleteReceipt: (slug: string, registrationId: number, receiptId: number) =>
    del<void>(
      `/admin/courses/${normalizeCourseSlug(slug)}/registrations/${requirePositiveInteger(registrationId, 'registrationId')}/receipts/${requirePositiveInteger(receiptId, 'receiptId')}`,
    ),
  createFollowUp: (slug: string, registrationId: number, payload: CourseRegistrationFollowUpCreate) =>
    post<CourseRegistrationFollowUpDTO>(
      `/admin/courses/${normalizeCourseSlug(slug)}/registrations/${requirePositiveInteger(registrationId, 'registrationId')}/follow-ups`,
      payload,
    ),
  updateFollowUp: (
    slug: string,
    registrationId: number,
    followUpId: number,
    payload: CourseRegistrationFollowUpUpdate,
  ) =>
    patch<CourseRegistrationFollowUpDTO>(
      `/admin/courses/${normalizeCourseSlug(slug)}/registrations/${requirePositiveInteger(registrationId, 'registrationId')}/follow-ups/${requirePositiveInteger(followUpId, 'followUpId')}`,
      payload,
    ),
  deleteFollowUp: (slug: string, registrationId: number, followUpId: number) =>
    del<void>(
      `/admin/courses/${normalizeCourseSlug(slug)}/registrations/${requirePositiveInteger(registrationId, 'registrationId')}/follow-ups/${requirePositiveInteger(followUpId, 'followUpId')}`,
    ),
  listRegistrations: (params?: { slug?: string; status?: string; limit?: number }) => {
    const search = new URLSearchParams();
    setTrimmedParam(search, 'slug', params?.slug);
    setTrimmedParam(search, 'status', params?.status);
    if (params?.limit != null) search.set('limit', String(requirePositiveInteger(params.limit, 'limit')));
    const qs = search.toString();
    return get<CourseRegistrationDTO[]>(`/admin/courses/registrations${qs ? `?${qs}` : ''}`);
  },
  listRegistrationEmails: (registrationId: number, limit = 100) =>
    get<CourseEmailEventDTO[]>(
      `/admin/courses/registrations/${requirePositiveInteger(registrationId, 'registrationId')}/emails?limit=${requirePositiveInteger(limit, 'limit')}`,
    ),
  getRegistration: (slug: string, registrationId: number) =>
    get<CourseRegistrationDTO>(
      `/admin/courses/${normalizeCourseSlug(slug)}/registrations/${requirePositiveInteger(registrationId, 'registrationId')}`,
    ),
};

File: tdf-hq/src/TDF/Routes/Courses.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE OverloadedStrings #-}

module TDF.Routes.Courses
  ( CourseSession(..)
  , SyllabusItem(..)
  , CourseMetadata(..)
  , UTMTags(..)
  , CourseRegistrationRequest(..)
  , CourseRegistrationResponse(..)
  , CourseRegistrationStatusUpdate(..)
  , CourseRegistrationNotesUpdate(..)
  , CourseRegistrationReceiptCreate(..)
  , CourseRegistrationReceiptUpdate(..)
  , CourseRegistrationFollowUpCreate(..)
  , CourseRegistrationFollowUpUpdate(..)
  , CourseUpsert(..)
  , CourseSessionIn(..)
  , CourseSyllabusIn(..)
  , CoursesPublicAPI
  , CoursesAdminAPI
  , WhatsAppHooksAPI
  , WhatsAppWebhookAPI
  ) where

import           Data.Aeson (FromJSON(parseJSON), Options(..), ToJSON, defaultOptions, genericParseJSON)
import           Data.Int (Int64)
import           Data.Text (Text)
import           Data.Time (Day)
import           GHC.Generics (Generic)
import           Servant

import           TDF.WhatsApp.Types (WAMetaWebhook)
import qualified TDF.DTO

data CourseSession = CourseSession
  { label :: Text
  , date  :: Day
  } deriving (Show, Generic)

instance ToJSON CourseSession
instance FromJSON CourseSession

data SyllabusItem = SyllabusItem
  { title  :: Text
  , topics :: [Text]
  } deriving (Show, Generic)

instance ToJSON SyllabusItem
instance FromJSON SyllabusItem

data CourseMetadata = CourseMetadata
  { slug           :: Text
  , title          :: Text
  , subtitle       :: Text
  , format         :: Text
  , duration       :: Text
  , price          :: Double
  , currency       :: Text
  , capacity       :: Int
  , remaining      :: Int
  , sessionStartHour :: Int
  , sessionDurationHours :: Int
  , locationLabel  :: Text
  , locationMapUrl :: Text
  , daws           :: [Text]
  , includes       :: [Text]
  , instructorName :: Maybe Text
  , instructorBio  :: Maybe Text
  , instructorAvatarUrl :: Maybe Text
  , sessions       :: [CourseSession]
  , syllabus       :: [SyllabusItem]
  , whatsappCtaUrl :: Text
  , landingUrl     :: Text
  } deriving (Show, Generic)

instance ToJSON CourseMetadata
instance FromJSON CourseMetadata

data UTMTags = UTMTags
  { source   :: Maybe Text
  , medium   :: Maybe Text
  , campaign :: Maybe Text
  , content  :: Maybe Text
  } deriving (Show, Generic)

instance FromJSON UTMTags where
  parseJSON = genericParseJSON strictObjectOptions
instance ToJSON UTMTags

data CourseRegistrationRequest = CourseRegistrationRequest
  { fullName  :: Maybe Text
  , email     :: Maybe Text
  , phoneE164 :: Maybe Text
  , source    :: Text
  , howHeard  :: Maybe Text
  , utm       :: Maybe UTMTags
  } deriving (Show, Generic)

instance FromJSON CourseRegistrationRequest where
  parseJSON = genericParseJSON strictObjectOptions
instance ToJSON CourseRegistrationRequest

data CourseRegistrationResponse = CourseRegistrationResponse
  { id     :: Int64
  , status :: Text
  } deriving (Show, Generic)

instance ToJSON CourseRegistrationResponse

data CourseRegistrationStatusUpdate = CourseRegistrationStatusUpdate
  { status :: Text
  } deriving (Show, Generic)

instance FromJSON CourseRegistrationStatusUpdate
instance ToJSON CourseRegistrationStatusUpdate

data CourseRegistrationNotesUpdate = CourseRegistrationNotesUpdate
  { notes :: Maybe Text
  } deriving (Show, Generic)

instance FromJSON CourseRegistrationNotesUpdate
instance ToJSON CourseRegistrationNotesUpdate

data CourseRegistrationReceiptCreate = CourseRegistrationReceiptCreate
  { fileUrl  :: Text
  , fileName :: Maybe Text
  , mimeType :: Maybe Text
  , notes    :: Maybe Text
  } deriving (Show, Generic)

instance FromJSON CourseRegistrationReceiptCreate
instance ToJSON CourseRegistrationReceiptCreate

data CourseRegistrationReceiptUpdate = CourseRegistrationReceiptUpdate
  { fileUrl  :: Maybe Text
  , fileName :: Maybe Text
  , mimeType :: Maybe Text
  , notes    :: Maybe Text
  } deriving (Show, Generic)

instance FromJSON CourseRegistrationReceiptUpdate
instance ToJSON CourseRegistrationReceiptUpdate

data CourseRegistrationFollowUpCreate = CourseRegistrationFollowUpCreate
  { entryType      :: Maybe Text
  , subject        :: Maybe Text
  , notes          :: Text
  , attachmentUrl  :: Maybe Text
  , attachmentName :: Maybe Text
  , nextFollowUpAt :: Maybe Text
  } deriving (Show, Generic)

instance FromJSON CourseRegistrationFollowUpCreate
instance ToJSON CourseRegistrationFollowUpCreate

data CourseRegistrationFollowUpUpdate = CourseRegistrationFollowUpUpdate
  { entryType      :: Maybe Text
  , subject        :: Maybe Text
  , notes          :: Maybe Text
  , attachmentUrl  :: Maybe Text
  , attachmentName :: Maybe Text
  , nextFollowUpAt :: Maybe Text
  } deriving (Show, Generic)

instance FromJSON CourseRegistrationFollowUpUpdate
instance ToJSON CourseRegistrationFollowUpUpdate

data CourseSessionIn = CourseSessionIn
  { label :: Text
  , date  :: Day
  , order :: Maybe Int
  } deriving (Show, Generic)
instance FromJSON CourseSessionIn
instance ToJSON CourseSessionIn

data CourseSyllabusIn = CourseSyllabusIn
  { title  :: Text
  , topics :: [Text]
  , order  :: Maybe Int
  } deriving (Show, Generic)
instance FromJSON CourseSyllabusIn
instance ToJSON CourseSyllabusIn

data CourseUpsert = CourseUpsert
  { slug                 :: Text
  , title                :: Text
  , subtitle             :: Maybe Text
  , format               :: Maybe Text
  , duration             :: Maybe Text
  , priceCents           :: Int
  , currency             :: Text
  , capacity             :: Int
  , sessionStartHour     :: Maybe Int
  , sessionDurationHours :: Maybe Int
  , locationLabel        :: Maybe Text
  , locationMapUrl       :: Maybe Text
  , whatsappCtaUrl       :: Maybe Text
  , landingUrl           :: Maybe Text
  , daws                 :: [Text]
  , includes             :: [Text]
  , instructorName       :: Maybe Text
  , instructorBio        :: Maybe Text
  , instructorAvatarUrl  :: Maybe Text
  , sessions             :: [CourseSessionIn]
  , syllabus             :: [CourseSyllabusIn]
  } deriving (Show, Generic)
instance FromJSON CourseUpsert
instance ToJSON CourseUpsert

type CoursesPublicAPI =
       "public" :> "courses" :> Capture "slug" Text :> Get '[JSON] CourseMetadata
  :<|> "public" :> "courses" :> Capture "slug" Text :> "registrations" :> ReqBody '[JSON] CourseRegistrationRequest :> PostCreated '[JSON] CourseRegistrationResponse

type CoursesAdminAPI =
       "courses" :> ReqBody '[JSON] CourseUpsert :> Post '[JSON] CourseMetadata
  :<|> "courses" :> "cohorts" :> Get '[JSON] [TDF.DTO.CourseCohortOptionDTO]
  :<|> "courses" :> "registrations" :>
         QueryParam "slug" Text :>
         QueryParam "status" Text :>
         QueryParam "limit" Int :>
         Get '[JSON] [TDF.DTO.CourseRegistrationDTO]
  :<|> "courses" :> Capture "slug" Text :> "registrations" :> Capture "registrationId" Int64 :> Get '[JSON] TDF.DTO.CourseRegistrationDTO
  :<|> "courses" :> Capture "slug" Text :> "registrations" :> Capture "registrationId" Int64 :> "dossier" :> Get '[JSON] TDF.DTO.CourseRegistrationDossierDTO
  :<|> "courses" :> "registrations" :> Capture "registrationId" Int64 :> "emails" :> QueryParam "limit" Int :> Get '[JSON] [TDF.DTO.CourseEmailEventDTO]
  :<|> "courses" :> Capture "slug" Text :> "registrations" :> Capture "registrationId" Int64 :> "status" :> ReqBody '[JSON] CourseRegistrationStatusUpdate :> Patch '[JSON] CourseRegistrationResponse
  :<|> "courses" :> Capture "slug" Text :> "registrations" :> Capture "registrationId" Int64 :> "notes" :> ReqBody '[JSON] CourseRegistrationNotesUpdate :> Patch '[JSON] TDF.DTO.CourseRegistrationDTO
  :<|> "courses" :> Capture "slug" Text :> "registrations" :> Capture "registrationId" Int64 :> "receipts" :> ReqBody '[JSON] CourseRegistrationReceiptCreate :> PostCreated '[JSON] TDF.DTO.CourseRegistrationReceiptDTO
  :<|> "courses" :> Capture "slug" Text :> "registrations" :> Capture "registrationId" Int64 :> "receipts" :> Capture "receiptId" Int64 :> ReqBody '[JSON] CourseRegistrationReceiptUpdate :> Patch '[JSON] TDF.DTO.CourseRegistrationReceiptDTO
  :<|> "courses" :> Capture "slug" Text :> "registrations" :> Capture "registrationId" Int64 :> "receipts" :> Capture "receiptId" Int64 :> Delete '[JSON] NoContent
  :<|> "courses" :> Capture "slug" Text :> "registrations" :> Capture "registrationId" Int64 :> "follow-ups" :> ReqBody '[JSON] CourseRegistrationFollowUpCreate :> PostCreated '[JSON] TDF.DTO.CourseRegistrationFollowUpDTO
  :<|> "courses" :> Capture "slug" Text :> "registrations" :> Capture "registrationId" Int64 :> "follow-ups" :> Capture "followUpId" Int64 :> ReqBody '[JSON] CourseRegistrationFollowUpUpdate :> Patch '[JSON] TDF.DTO.CourseRegistrationFollowUpDTO
  :<|> "courses" :> Capture "slug" Text :> "registrations" :> Capture "registrationId" Int64 :> "follow-ups" :> Capture "followUpId" Int64 :> Delete '[JSON] NoContent

type WhatsAppWebhookAPI =
       "webhooks" :> "whatsapp" :> QueryParam "hub.mode" Text :> QueryParam "hub.verify_token" Text :> QueryParam "hub.challenge" Text :> Get '[PlainText] Text
  :<|> "webhooks" :> "whatsapp" :> ReqBody '[JSON] WAMetaWebhook :> Post '[JSON] NoContent

type WhatsAppHooksAPI =
       "hooks" :> "whatsapp" :> QueryParam "hub.mode" Text :> QueryParam "hub.verify_token" Text :> QueryParam "hub.challenge" Text :> Get '[PlainText] Text
  :<|> "hooks" :> "whatsapp" :> ReqBody '[JSON] WAMetaWebhook :> Post '[JSON] NoContent

strictObjectOptions :: Options
strictObjectOptions = defaultOptions { rejectUnknownFields = True }
